guava/src/module-info.java

```diff
@@ -0,0 +1,45 @@
+// GENERATED FILE - DO NOT EDIT
+
+/*
+ * Copyright (C) 2008 The Guava Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Google Guava
+ */
+module com.google.common {
+  requires java.base;
+  requires java.logging;
+  requires jdk.unsupported;
```

Comment:
Curious as to which parts of Guava require `jdk.unsupported`?
Not sure if it's correct to explicitly add a module requirement on it, since this is not a module dictated in the JLS, so it could be changed/removed/altered in some incompatible way in future Java releases?

Reply:
@norrisjeremy I believe this is because of `sun.misc.Unsafe`; I'm not sure there is any way around `requires jdk.unsupported` at least for the time being. Once Unsafe is refactored away this can be removed. We're 10 years in to the launch of JPMS and `jdk.unsupported` is still around, and it is supported on all current Java versions up to latest LTS, so it should be okay.

Reply:
Ok, I suppose.
But not sure if you have seen [JDK-8323072](https://bugs.openjdk.org/browse/JDK-8323072) & it's associated [draft JEP](https://openjdk.org/jeps/8323072), which I believe may land in Java 23 later this year.
I think the doomsday clock for `sun.misc.Unsafe` is advancing ever closer to midnight.

Reply:
@norrisjeremy Oh, agreed, for sure; but this PR just focused on JPMS for now. JPMS and Unsafe are both sensitive changes that should be done soon, no question.

Java 22 comes out this month, in about 10 days. So you're right to be concerned for Java 23, but Guava has some time to address that before it becomes a problem. JPMS, meanwhile, has been out since Java 9 and is _currently_ unsupported by Guava, so this PR advances on that front.

Even if Unsafe is deprecated in Java 23 (I don't foresee it getting removed completely yet), this changeset will let Guava work at Java 9 - Java 22 with fully modular Java. Then, working to refactor/fix the `Unsafe` usages becomes easier, because Guava is shipping MRJARs, so `Unsafe` can safely be used for earlier JVMs only, with later JVMs providing classes that override with other intrinsics.

Reply:
Yes, that makes sense.
Btw, finally adding JPMS support is greatly appreciated btw. :)

Reply:
@norrisjeremy Thank you likewise for the reviews :)

Reply:
Does guava require unsafe, or can it safely fallback?

Reply:
@Sineaggi If I understand correctly, the implications of that are basically the modular equivalent of a compile-time dependency, right? So users could effectively opt-in by adding `requires jdk.unsupported`?

In any case, great catch. You're right: the testsuite passes in full with only `requires static`.

@cpovirk Curious about your thoughts on behavior defaults here. By the way: _technically_ it **is legal** to have a slightly different `module-info.java` for different versions of a single JAR. The public API surface of the module graph can't change, but including things from `java` or `jdk` was specifically imagined to be possible, so that newer class suites can leverage the latest JVM features.

That should mean that it is possible, in theory, to build with `requires` or `requires static` for `jdk.unsupported` based on the active JVM. It would add some build complexity, but it's possible. So I think that means the options are:

- **`requires jdk.unsupported`:** All users are opting in to `Unsafe` whether they like it or not; this is a drawback but the benefit is that the library performs as expected out of the box (i.e. it implies no performance hit from falling back from `Unsafe`)

- **`requires static jdk.unsupported`:** All users can opt-in to use of `Unsafe`; when they upgrade Guava, depending on the shape of their workload, they may see a performance hit, but _only_ if they are using fully modular Java, which would be a new use case at that library version anyway. Otherwise, on the classpath, there is no isolation and so there is no difference.

- **Varied by JVM level:** Perhaps, for users of Java 9 or even Java 11, we use `requires jdk.unsupported`, so that the library performs as expected for the vast "legacy" use cases, which happen to begin building with Guava on the `modulepath` simply because it ships a `module-info.java`. For Java 11+, the `module-info.java` specifies `requires static jdk.unsupported`, and we trust more modern Java users (who are building in a fully modular manner, remember, classpath is still fine) to add `requires jdk.unsupported` if they need it.

Thoughts? Maybe this is overthinking it, I'm not sure.

Reply:
(Also, it would be helpful to see the actual performance hit that should realistically be expected; I'm midway through breaking the benchmark suite out and modernizing with JMH, so perhaps we'll look to some numbers from there or Calibre to help inform this. I think a big question mark is exactly _what_ performance degradation loss would be expected with the loss of `Unsafe` as a default.)

Reply:
Yes, perhaps adding `requires static jdk.unsupported` to avoid a hard requirement on it at runtime would work?
I believe users that are performance sensitive and also desiring the ability to then also use Guava on the module-path instead of class-path could still make it work by adding `--add-exports jdk.unsupported/sun.misc=com.google.common --add-opens jdk.unsupported/sun.misc=com.google.common` to their application.